doubly_linked_list.py

```diff
@@ -3,58 +3,107 @@
 
 class DoublyLinkedList:
     def __init__(self):
-        self.head = None
+        self.head = Node(None, None)
+        self.tail = Node(None, None, prev_node=self.head)
+        self.head.next_node = self.tail
+        self.node_count = 0
 
     def fetch_head(self):
         return self.head
 
     def fetch_tail(self):
-        temp = self.head
-        while temp.next_node:
-            temp = temp.next_node
-        return temp
+        # TC : O(1)
+        return self.tail
+
+    def add(self, value):
+        #TC : O(1)
+        if not value:
+            return False
+
+        head = self.fetch_head()
+        tail = self.fetch_tail()
 
-    def add_node(self, value):
-        if self.head is None:
-            self.head = Node(key=None, value=value, next_node=None, prev_node=None)
+        if self.node_count == 0:
+            new_node = Node(key=None, value=value, next_node=None, prev_node=None)
+            new_node.next_node = head.next_node
+            new_node.prev_node = head.next_node.prev_node
```

Comment:
With sentinel node approach... the  part will be
```
head.next = tail
tail.prev = head
new_node.prev_node = head.next_node.prev_node == tail.prev === head
new_node.prev_node = head

```

Reply:
handled